Panels, Columns, Rows

thisisdano opened the pull request:
**Work in Progress**
**Pair with https://github.com/18F/cg-dashboard/pull/929**

thisisdano committed:
Merge branch 'master' into dw-panel-style

thisisdano committed:
styles for panel-secondary

thisisdano committed:
styling for secondary link

thisisdano committed:
rebuild usage panel with new components

thisisdano committed:
make raw div slurp parent row

thisisdano committed:
cleanup routes rows and buttons

thisisdano committed:
implemented 2-col routes and services

thisisdano committed:
first pass at services panel default

thisisdano committed:
style modifications for dir tree

thisisdano committed:
fixed lint error

thisisdano committed:
new card styles

msecret commented on src/css/components/panel.scss:
```diff
@@ -29,30 +28,32 @@ $panel-padding: $grid-3;
+  margin-bottom: $grid-5;
 }
 
 .panel-secondary {
```
What is panel secondary? What's it's purpose? I think defining that and naming it so it matches it's purpose might make it easier to use.

line47 replied:
Correct me if I'm wrong @thisisdano but I believe that is for a panel subtitle

thisisdano replied:
This is for the information that goes at the top-right of the panel.

thisisdano replied:
i.e. the logs message in the following panel:

![screen shot 2017-01-06 at 12 29 01 pm](https://cloud.githubusercontent.com/assets/11464021/21731920/b3e8ef88-d40b-11e6-82d4-8d1f4f8d15be.png)

thisisdano replied:
I'm happy to use a different name for the class if there are suggestions.

msecret commented on src/css/components/panel.scss:
```diff
@@ -68,61 +69,44 @@ h4 + .panel {
-    &:last-child {
-      margin-right: 0;
-    }
+  .action {
```
Is there a way we can do this without nesting actions? There may not be and that's fine, just wanted to point out.

thisisdano replied:
We could possibly do this with even margins and negative margin on the container (I like this), but this would just prevent the necessity of the `.panel-actions-right .action` nest. I don't _think_ you can apply the margin globally to action since the actions tend to appear in a wide-ish range of locations. This could be regularized, but it isn't regular yet.

thisisdano replied:
Oops, nope. I tried this and the negative margin was great for the actions, but broke adjacent columns in a way I couldn't fix.

msecret commented on src/css/components/panel.scss:
```diff
@@ -132,22 +116,16 @@ h4 + .panel {
 }
 
-.panel-row-boxed {
+.card {
```
Is there a reason this is here rather then in card component?

thisisdano replied:
oh huh — I thought I moved it.

thisisdano replied:
ahh, moved it but never deleted it from `panel`

thisisdano replied:
fixed

msecret commented on src/css/components/panel.scss:
```diff
@@ -252,3 +230,14 @@ h4 + .panel {
+
+// a couple random panel styles
+
+.route-link {
```
Similar comment to `card`, I'd prefer to keep only things related to `panel` in here so everything is easier to find and it keeps the current organization consistent. Is there a reason it should be in here though?

thisisdano replied:
Agreed. I'd only done it because there weren't either `routes.scss` or `service-instances.scss` files, and I was hesitant to add them with only a single rule. This was kind of a kludge.

Should I just go ahead and add those two files?

msecret commented on src/css/components/card.scss:
```diff
@@ -7,7 +7,20 @@
+    padding-bottom: $grid-2;
+  }
+
+  .panel-row:last-child {
```
I like how you used `>` in the first rule above, can that be done here as well to keep more specific?

Commit:
Fixing button alignment for editing route

line47 commented on src/css/components/panel.scss:
```diff
@@ -233,6 +227,18 @@ $panel-padding: $grid-3;
+      }
+    }
+}
+
```
It was doing this:
![screen shot 2017-01-06 at 2 45 41 pm](https://cloud.githubusercontent.com/assets/1449852/21730697/dacd5720-d41e-11e6-94c8-2300491a80ce.png)

After this fix the buttons are aligned right and don't have as much space between the buttons.

thisisdano replied:
This panel is one of the next on my list!

msecret reviewed (commented):
Things are generally looking good, the thing I want to think about more is row, col and how you ahve to set start and end rows. Looking at the PR in dashboard, some stuff looks a lot more complicated to build now. Before you could just drop some panels, and now there's a lot more components and props you have to use to get the layout. I'm not saying that's bad, I just want to think about the component interface for developers.

I'm a fan of the card component though. I might try and find time to look at the col, row components and get a sense of how I'd use them and where things can be simplified.

Commit:
Fixing lint issues

thisisdano committed:
removed card code from panel

thisisdano committed:
fixed last-child card padding

thisisdano committed:
Merge branch 'dw-panel-style' of github.com:18F/cg-style into dw-panel-style

thisisdano committed:
split out service-instance rule

thisisdano committed:
split out service-instance rule

thisisdano commented:
@msecret 
You don't have to use start and end rows — those are alignment props, for setting the vertical alignment of the row (i.e. `flex-start`, `center`, and `flex-end`). Rows default to middle (I guess this should be `center`).

thisisdano commented:
I truly think this is a simpler system for devs — columns and rows are pretty straightforward for layouts, many things work well in the default state, and a few simple props open up more complex layout options. (Perhaps obviously) I personally find the new system to be simpler, cleaner, easier to read, and more predictable — tho ymmv.

I _am_ actively looking to improve this. :)

thisisdano committed:
added valign values to rows

thisisdano commented:
Clarified those alignment props to be `valign` and used `top` `center` and `bottom` as values

thisisdano committed:
Merge branch 'master' into dw-panel-style

adborden commented on docs/src/components/features/Tree View/space-tree-view.html:
```diff
@@ -10,7 +10,7 @@ <h3 class="contents-secondary">
         </span>
         <span class="panel-column panel-column-less">
-        {% render '@status--running' %}
+        {% render '@status--ok' %}
```
Oops, didn't realize these were here. Thanks for fixing!

thisisdano committed:
cleanup contents-tree rows

adborden reviewed (commented):
I'm a fan of having the row/col with clear rules. I'd caution about where panel is going, it seems like it's one component to rule them all. It handles some content related aspects like header/content/footer/actions, as well as presentation plain/boxed/lined. I could see us breaking down the panel pieces into smaller composable chunks.

thisisdano commented:
@adborden I'd agree that this leaves room to further component-ize special cases — `PanelRow` for instance instead of a `type`, and perhaps others. For now, this felt simpler _to me_, but I'm certainly open to more splitting and more components.

thisisdano committed:
regularized contents tree; first pass at docs

thisisdano committed:
first pass at contents tree docs

thisisdano committed:
cleaned up tree docs templates

thisisdano committed:
panel-row > panel-entry

thisisdano committed:
updated panel-entry references in docs

thisisdano committed:
removed unnecessary card > panel-entry padding rule

thisisdano committed:
removed path from contents-tree-app in PL

thisisdano committed:
replaced circle spinner with dots loading

thisisdano committed:
removed loading.gif

thisisdano committed:
some fixes for loading indicator

thisisdano committed:
fixed main height position

thisisdano committed:
added text display for loading indicator

thisisdano committed:
cleanup ContentsTree PL docs

thisisdano committed:
secondary > helptertext

thisisdano committed:
resize card per no org icon

thisisdano committed:
componentize ContentsTreeColumn

thisisdano committed:
added first pass docs to Row and Col components

thisisdano committed:
added first pass docs to Row and Col components

thisisdano committed:
typos in Row and Col docs; bumped to beta

thisisdano committed:
more typos

thisisdano committed:
ffs another typo...

thisisdano committed:
documented new Loading in PL

thisisdano committed:
fixed lint

thisisdano commented:
OK — I've made a number of additions and improvements over the weekend:
- Simplified `Row`
  - Now only takes `valign` and `gutters` props
  - `panel-row` type now its own component: `PanelEntry` [not in PL yet]
- Simplified `Col`
  - new component `ContentsTreeColumn` for data columns in the contents tree [not in PL yet]
- Regularized contents tree components
  - `AppQuicklook` now `ContentsTreeApp` [documented in PL]
  - `SpaceQuicklook` now `ContentsTreeSpace` [documented in PL]
  - documented
- Changed block loading component
  - now the same as inline loading indicator
  - now can display loading text
  - made a regularized loading container (`.loading-container`) for loading in panels and cards. This prevents panel "jumping" on loading. [documented in PL]
- removed icon from Org in overview card
- removed the gray path from `ContentsTreeApp` [documented in PL]
- Changed `secondary` prop in `Card` (the top-right content) to `helperText` [not in PL yet]
- Gave `main` a min-height to keep the footer at the bottom of the window

![loading](https://cloud.githubusercontent.com/assets/11464021/21766740/14ea1fbc-d625-11e6-88a0-6e71f8e5d6ab.gif)

Changing loading revealed that expand actions in the overview trigger loads on _all_ open cards.

We also get the disconcerting "double-load" on the Overview as well, with a quick load (perhaps with no status information) that leads to a second, more complete load in quick succession. The preliminary load feels unnecessary and confusing.

thisisdano committed:
documented panel and variants in PL

thisisdano committed:
documented panel and variants in PL

thisisdano committed:
update space and org overview headers

thisisdano committed:
added Card to PL

thisisdano committed:
added PanelEntry docs in PL

thisisdano committed:
rearranged items in PL

thisisdano committed:
rearranged items in PL

thisisdano commented on docs/src/components/components/card/README.md:
```diff
@@ -0,0 +1,9 @@
+none
+
+### Defaults
+none
```
Documenting new Card component

thisisdano commented on docs/src/components/components/card/card.html:
```diff
@@ -1,9 +1,8 @@
 {% for org in organizations %}
-<div class="panel-row-boxed">
+<div class="card">
```
Using this pattern to show general Card rather than the Overview feature specifically.

thisisdano commented on docs/src/components/components/card/card.html:
```diff
@@ -1,9 +1,8 @@
     <div class="panel-row-is_clickable">
         <div class="panel-column">
             <h2 class="card-title-primary">
-        {% render '@icon-square--org-medium' %}
```
No longer including icon in this view — it wasn't clear to users

thisisdano commented on docs/src/components/components/contents tree/contents-tree-app.config.yml:
```diff
@@ -0,0 +1,15 @@
+title: ContentsTreeOrg (Contents tree / app contents)
```
Oops, this should be `ContentsTreeApp`

thisisdano committed:
fixed ContentsTreeApp typo

thisisdano committed:
fixed ContentsTreeApp code typo

thisisdano commented on docs/src/components/components/contents tree/contents-tree-app.html:
```diff
@@ -0,0 +1,33 @@
+{# ContentsTreeApp status="OK" #}
```
Oops. Can remove status from this note...

thisisdano commented on docs/src/components/components/loading/loading.html:
```diff
@@ -1,8 +1,10 @@
-  <img class="loading-indicator"
-      alt="Loading"
-      src="/img/loading.gif" />
+<div class="loading loading-container" role="alertdialog">
```
gotta add back the Aria info

thisisdano replied:
done

thisisdano committed:
added aria back into loading docs

thisisdano commented on docs/src/components/features/Tree View/space-tree-view.html:
```diff
@@ -1,45 +0,0 @@
-<div class="panel-row">
```
replaced with `contents-tree`

thisisdano commented on docs/src/components/features/panel/panel.html:
```diff
@@ -1,37 +1,40 @@
-<section class="panel">
-    <header class="panel-header row-center-x">
+<!-- `div` should be `section` -->
```
Brought this up to date w production code, but made a note where we should update production code semantics

thisisdano commented on docs/src/components/features/panel/panel.html:
```diff
@@ -1,37 +1,40 @@
+      </div>
     </div>
-</section>
+
```
brought `panel-actions` container up to date

thisisdano commented on docs/src/main.css:
```diff
@@ -271,7 +271,6 @@ body {
   overflow: hidden;
   background-color: #d3d3d3;
   height: 16px;
-  width: 100%;
```
just changes for PL display

thisisdano commented on src/css/base/background.scss:
```diff
@@ -8,5 +8,6 @@ body {
 
 main {
   background-color: transparent;
+  min-height: calc(100% - 22rem);
```
this keeps the footer at the bottom of the page

thisisdano commented on src/css/base/global.scss:
```diff
@@ -7,7 +7,3 @@ body {
   font-size: 14px;
   margin: 0;
 }
-
```
dunno why we had this. it was wrong

thisisdano commented on src/css/components/all.scss:
```diff
@@ -29,6 +29,7 @@
 @import 'sidenav.scss';
 @import 'search.scss';
 @import 'section.scss';
+@import 'service-instance.scss';
```
needed for a couple service instance panel styles

thisisdano commented on src/css/components/card.scss:
```diff
@@ -7,7 +7,15 @@
   border-radius: $rounded;
   margin-bottom: $grid-2;
-  padding: 0;
+  padding: $grid-3 $grid-2;
```
regularized and passed flex on to the clickable div

thisisdano commented on src/css/components/contents.scss:
```diff
@@ -21,3 +22,67 @@
 .contents-path {
   color: $color-mediumgray;
 }
+
```
All the new `contents-tree` styles

thisisdano commented on src/css/components/contents.scss:
```diff
@@ -21,3 +22,67 @@
+  width: 100%;
+}
+
+.contents-tree-space {
```
gets a little complicated to control nested border colors...

thisisdano commented on src/css/components/grid.scss:
```diff
@@ -32,3 +32,77 @@
+  align-items: center;
+  flex-wrap: wrap;
+
+  // hack for react containing div requirement
```
these rules pass flex values on to unclassed child `div`s which are sometimes necessary in React templates

thisisdano commented on src/css/components/grid.scss:
```diff
@@ -32,3 +32,77 @@
+}
+
+$gutter: $grid-4;
+$gutter-offset: $grid-2; // $gutter / 2
```
I haven't had good luck getting math operators to work well in the sass — keep throwing errors — so I hardcoded

thisisdano commented on src/css/components/loading.scss:
```diff
@@ -2,25 +2,66 @@
 $z-foreground: 200;
 
 .loading {
```
Absolutely positioned the block loader to get it to properly display full height of parent

thisisdano commented on src/css/components/loading.scss:
```diff
@@ -2,25 +2,66 @@
   z-index: $z-foreground;
 }
 
+.loading-container {
```
this is a container for the block loader if it is nested in something like a panel

thisisdano commented on src/css/components/loading.scss:
```diff
@@ -2,25 +2,66 @@
 }
 
-$indicator-size: 3.5rem;
+.loading-information {
```
the new general container for the block loading indicator — sized so it can accommodate text

thisisdano commented on src/css/components/loading.scss:
```diff
@@ -2,25 +2,66 @@
-  position: absolute;
-  top: calc(50% - #{$indicator-size});
-  width: $indicator-size;
+
```
a bigger loading indicator for the block loader

thisisdano commented on src/css/components/panel.scss:
```diff
@@ -9,7 +9,6 @@ $panel-padding: $grid-3;
   @include clearfix;
   background-color: $color-white;
   margin-bottom: $grid-4;
-  // TODO make padding and margins consistent
```
TODID it

thisisdano commented on src/css/components/panel.scss:
```diff
@@ -29,30 +28,32 @@ $panel-padding: $grid-3;
 
 }
 
-// TODO this is a hack, panel-row borders need to be reconciled, panel-row vs
```
TODID this

thisisdano commented on src/css/components/panel.scss:
```diff
@@ -29,30 +28,32 @@ $panel-padding: $grid-3;
 }
 
-.panel-secondary {
+.panel-helpertext {
```
This is the new class for the helper text that appears in the top-right of a panel

thisisdano commented on src/css/components/panel.scss:
```diff
@@ -68,61 +69,44 @@ h4 + .panel {
 
-    & + .action {
-      margin-left: .5rem;
+    .action {
```
made this a regular margin so no need for adjacency selector

thisisdano commented on src/css/components/panel.scss:
```diff
@@ -68,61 +69,44 @@ h4 + .panel {
-.panel-row {
-  $vertical-padding: $grid-105;
-  border-top: 1px solid $color-textblack;
+.panel-entry {
```
`panel-row` is now `panel-entry`

thisisdano commented on src/css/components/panel.scss:
```diff
@@ -247,8 +177,22 @@ h4 + .panel {
   padding: $grid-3 $grid-3 $grid-4;
 }
+
+// a couple random panel styles
```
gotta see if this is still necessary...

thisisdano commented on src/css/components/panel.scss:
```diff
@@ -132,54 +116,19 @@ h4 + .panel {
     width: 100%;
   }
 
-  > .panel-row {
```
Amost all of these variations have been consolidated in to the `Row` component or into `PanelEntry` or `Card`

thisisdano commented on src/css/components/stat.scss:
```diff
@@ -57,21 +57,38 @@ h5 + .stat {
-  background-color: $color-lightestgray;
-  margin: -$v-space 1.8rem;
-  padding: $v-space 0;
+  background-color: $color-cream;
```
All changes to make the Instances box display properly

thisisdano commented on src/css/components/stat.scss:
```diff
@@ -57,21 +57,38 @@ h5 + .stat {
   text-align: center;
+  top: 0;
+
+  .subtext {
```
This may need an adjustment in the not-too-distant future...

thisisdano reviewed (commented):
My notes on these changes

thisisdano committed:
added style for row-stretch

thisisdano committed:
documented col-stretch prop

thisisdano committed:
fixed merge with master

thisisdano committed:
replaced errant icon-squares
